alf/utils/schedulers.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) 2020 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Schedulers."""
+from alf.trainers.policy_trainer import Trainer
+import gin
+
+
+class Scheduler(object):
+    """Base class of all schedulers.
+
+    A scheduler is used to generate manually defined values based on the training
+    progress.
+
+    The subclass should call ``progress()`` to get the current training progress
+    and use it to calculate the scheduled value. There are three types of training
+    progresses:
+
+    * "percent": percent of training completed.
+    * "iterations": the number training iterations.
+    * "env_steps": the number of environment steps
+
+    """
+
+    def __init__(self, progress_type):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+        """
+        if progress_type == "percent":
+            self._progress_func = Trainer.progress
+        elif progress_type == "iterations":
+            self._progress_func = Trainer.current_iterations
+        elif progress_type == "env_steps":
+            self._progress_func = Trainer.current_env_steps
+        else:
+            raise ValueError("Unknown progress_type: %s" % progress_type)
+
+    def progress(self):
+        return self._progress_func()
+
+
+class ConstantScheduler(object):
+    def __init__(self, value):
+        self._value = value
+
+    def __call__(self):
+        return self._value
+
+
+@gin.configurable
+class StepScheduler(Scheduler):
+    """There is one value for each defined region of training progress."""
+
+    def __init__(self, progress_type, schedule):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+            schedule (list[tuple]): each tuple is a pair of (progress, value)
+                which means that if current progress is less than progress,
+                the current value will be value.
+        """
+        super().__init__(progress_type)
+        self._progresses, self._values = zip(*schedule)
+        self._index = 0
+
+    def __call__(self):
+        progress = self.progress()
+        index = self._index
+        progresses = self._progresses
+        if progress < progresses[index]:
+            while index > 0 and progress < progresses[index - 1]:
```

Comment:
In what situation we need to backstep the index? It seems that the progress is always incrementing